hub.go

```diff
@@ -67,60 +70,98 @@ func (hub *Hub) LastEventID() EventID {
 }
 
 func (hub *Hub) stackTop() *layer {
+	hub.RLock()
 	stack := hub.stack
 	if stack == nil || len(*stack) == 0 {
+		hub.RUnlock()
 		return nil
 	}
-	return (*stack)[len(*stack)-1]
+	top := (*stack)[len(*stack)-1]
+	hub.RUnlock()
+
+	return top
 }
 
 // Clone returns a copy of the current Hub with top-most scope and client copied over.
 func (hub *Hub) Clone() *Hub {
-	return NewHub(hub.Client(), hub.Scope().Clone())
+	clone := NewHub(hub.Client(), hub.Scope().Clone())
```

Comment:
assuming this is semantically correct, wouldn't it make sense to have the Read lock here once and the clones done under the lock? instead of calling two methods that will individually lock?

In between getting the client and scope, things might have changed.